ext/bcmath/libbcmath/src/round.c

```diff
@@ -0,0 +1,164 @@
+/* round.c: bcmath library file. */
+/*
+    Copyright (C) 1991, 1992, 1993, 1994, 1997 Free Software Foundation, Inc.
+    Copyright (C) 2000 Philip A. Nelson
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.  (LICENSE)
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to:
+
+      The Free Software Foundation, Inc.
+      59 Temple Place, Suite 330
+      Boston, MA 02111-1307 USA.
+
+    You may contact the author by:
+       e-mail:  philnelson@acm.org
+      us-mail:  Philip A. Nelson
+                Computer Science Department, 9062
+                Western Washington University
+                Bellingham, WA 98226-9062
+
+*************************************************************************/
+
+#include "bcmath.h"
+#include "private.h"
+#include <stddef.h>
+
+void bc_round(bc_num num, zend_long precision, zend_long mode, bc_num *result)
+{
+	/* clear result */
+	bc_free_num(result);
+
+	/*
+	* The following three cases result in an early return:
+	*
+	* 1. When rounding to an integer part, when trying to round to a digit
+	*    larger than the num.
+	* 2. If try to round to the finest digit of num or finer than that,
+	*    num is returned unchanged.
+	* 3. If the fraction part is all 0 from the beginning or middle to the
+	*    end, the 0's are omitted and the number of digits in num is reduced.
+	*    In that case, may end up in the same situation as 2.
+	*/
+	if ((int) num->n_len + precision <= 0) {
```

Comment:
Oh, that's true! I'll fix it.

Reply:
Oh? Is it perhaps wrong to receive precision as an int?

Reply:
In PHP we have the int type, but in C this maps to zend_long. zend_long and int, in C, are only the same thing on 32-bit, but on 64-bit zend_long is wider than int (64bit zend_long vs 32bit int).
So it's not wrong to have the parameter type of `precision` be a PHP int.
`num->n_len` is a 64-bit unsigned value because its type is `size_t`, so you shouldn't downcast.

Reply:
Thank you for the detailed explanation!

I was thinking that if the precision is zend_long, the precision can only be specified up to about half the maximum number of digits of n_scale due to the difference in the presence or absence of a sign. (I don't think there is actually a demand for such a use case...)

Reply:
@nielsdos On 64 bit systems, PHP's int is also 64 bit long (https://3v4l.org/VEkMt#v8.3.1). Or did I understand your explanation wrong?

Reply:
@kocsismate I meant C's int is still 32 bits.

Reply:
It was completely my mistake to downcast to `int`. Please pretend that C's int did not exist here.....

Actually I have to think about `size_t` for `n_scale` and `zend_long` for `precision`.

Is `size_t` 64bit depending on the environment? In that case, the maximum value of `size_t` would be about twice the maximum value of `zend_long`, so I was concerned about that.

Reply:
> Is size_t 64bit depending on the environment?

Yes C's size_t is 64-bit on 64-bit systems, and 32-bit on 32-bit systems.